File: uilib/src/main/java/dev/entao/ui/util/LayerUtil.java
package dev.entao.ui.util;

import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import dev.entao.appbase.App;

import java.util.ArrayList;

public class LayerUtil {
	private static class TempDrawable {
		public Drawable drawable;
		public int left = 0;
		public int top = 0;
		public int right = 0;
		public int bottom = 0;

		public TempDrawable(Drawable d, int n) {
			drawable = d;
			left = top = right = bottom = n;
		}
	}

	ArrayList<TempDrawable> ls = new ArrayList<TempDrawable>();

	//inset dp
	public void add(Drawable d, int insetLeftDp, int insetTopDp, int insetRightDp, int insetBottomDp) {
		TempDrawable td = new TempDrawable(d, 0);
		td.left = insetLeftDp;
		td.top = insetTopDp;
		td.right = insetRightDp;
		td.bottom = insetBottomDp;
		ls.add(td);
	}

	public void add(Drawable d, int insetDp) {
		TempDrawable t = new TempDrawable(d, insetDp);
		ls.add(t);
	}

	public void add(Drawable d) {
		TempDrawable t = new TempDrawable(d, 0);
		ls.add(t);
	}

	public LayerDrawable get() {
		Drawable[] layers = new Drawable[ls.size()];
		for (int i = 0; i < ls.size(); ++i) {
			layers[i] = ls.get(i).drawable;
		}
		LayerDrawable ld = new LayerDrawable(layers);
		for (int i = 0; i < ls.size(); ++i) {
			TempDrawable t = ls.get(i);
			ld.setLayerInset(i, App.INSTANCE.dp2px(t.left), App.INSTANCE.dp2px(t.top), App.INSTANCE.dp2px(t.right), App.INSTANCE.dp2px(t.bottom));
		}
		return ld;
	}

	public static LayerDrawable build(Drawable d, int inset) {
		LayerDrawable ld = new LayerDrawable(new Drawable[]{d});
		ld.setLayerInset(0, App.INSTANCE.dp2px(inset), App.INSTANCE.dp2px(inset), App.INSTANCE.dp2px(inset), App.INSTANCE.dp2px(inset));
		return ld;
	}
}

File: uilib/src/main/java/dev/entao/kan/util/LayerUtil.kt
package dev.entao.kan.util

import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import dev.entao.kan.appbase.App
import java.util.*

class LayerUtil {

     var ls = ArrayList<TempDrawable>()

     class TempDrawable(var drawable: Drawable, n: Int) {
        var left = 0
        var top = 0
        var right = 0
        var bottom = 0

        init {
            bottom = n
            right = bottom
            top = right
            left = top
        }
    }

    //inset dp
    fun add(d: Drawable, insetLeftDp: Int, insetTopDp: Int, insetRightDp: Int, insetBottomDp: Int) {
        val td = TempDrawable(d, 0)
        td.left = insetLeftDp
        td.top = insetTopDp
        td.right = insetRightDp
        td.bottom = insetBottomDp
        ls.add(td)
    }

    fun add(d: Drawable, insetDp: Int) {
        val t = TempDrawable(d, insetDp)
        ls.add(t)
    }

    fun add(d: Drawable) {
        val t = TempDrawable(d, 0)
        ls.add(t)
    }

    fun get(): LayerDrawable {
        val layers = arrayOfNulls<Drawable>(ls.size)
        for (i in ls.indices) {
            layers[i] = ls[i].drawable
        }
        val ld = LayerDrawable(layers)
        for (i in ls.indices) {
            val t = ls[i]
            ld.setLayerInset(i, App.dp2px(t.left), App.dp2px(t.top), App.dp2px(t.right), App.dp2px(t.bottom))
        }
        return ld
    }

    companion object {

        fun build(d: Drawable, inset: Int): LayerDrawable {
            val ld = LayerDrawable(arrayOf(d))
            ld.setLayerInset(0, App.dp2px(inset), App.dp2px(inset), App.dp2px(inset), App.dp2px(inset))
            return ld
        }
    }
}
